File: 卡码/代码随想录/1/04-字符串/Java/src/_06_重复的子字符串/LeetCode_0459.java
//459. 重复的子字符串
package _06_重复的子字符串;

public class LeetCode_0459 {

}
class Solution {
  public boolean repeatedSubstringPattern(String s) {
  	int n=s.length();
  	//法一
//  	String s1=(s+s).substring(1,2*n-1);
//  	return s1.contains(s);
  	//法二
  	int[] next=new int[n];
  	int j=0;
  	next[0]=0;
  	for(int i=1;i<s.length();i++) {
  		while(j>0&&s.charAt(j)!=s.charAt(i)) j=next[j-1];
  		if(s.charAt(j)==s.charAt(i)) j++;
  		next[i]=j;
  	}
  	return next[n-1]!=0&&n%(n-next[n-1])==0;	//最长相等前后缀不等于0且剩下的元素能被字符串长度整除则一定有重复子串构成
  }
}

File: 卡码/代码随想录/Go/字符串/重复的子字符串/LeetCode_0459.go
// 459. 重复的子字符串
package main

import "strings"

func repeatedSubstringPattern(s string) bool {
	n:=len(s)
	//法一
	// s1:=s+s
	// return strings.Contains(s1[1:2*n-1],s)
	//法二
	next:=make([]int,n)
	j:=0
	next[0]=0
	for i:=1;i<n;i++ {
		for j>0&&s[i]!=s[j] {
			j=next[j-1]
		}
		if s[i]==s[j] {
			j++
		}
		next[i]=j
	}
	return next[n-1]!=0&&n%(n-next[n-1])==0
}
